### PR TITLE
Introduce a fake clock to F3

### PR DESCRIPTION
**Pull request by Kubuxu**

### CODE REVIEW

**Comment by Kubuxu** [comment]
no idea why some tests  are still slow-ish. They seem to send a lot of messages

File: certexchange/polling/common_test.go [hunk 1]
```diff
@@ -9,19 +9,14 @@ import (
...
 // The network name used in tests.
 const TestNetworkName gpbft.NetworkName = "testnet"
 
 // The clock used in tests. Time doesn't pass in tests unless you add time to this clock.
-var MockClock = clock.NewMock()
-
-func init() {
-	clk = MockClock
-}
+//var MockClock = clock.NewMock()
```
Comment (Stebalien) [inline certexchange/polling/common_test.go:19]: removeme

File: internal/clock/clock.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+type clockKeyType struct{}
+
+var clockKey = clockKeyType{}
+
+func NewMock() *Mock {
+	return clock.NewMock()
+}
+
+func WithMockClock(ctx context.Context) (context.Context, *Mock) {
```
Comment (Stebalien) [inline internal/clock/clock.go:21]: Docs would be nice.

File: internal/clock/clock.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+type Clock = clock.Clock
+type Mock = clock.Mock
+type Timer = clock.Timer
+
+type clockKeyType struct{}
+
+var clockKey = clockKeyType{}
+
+func NewMock() *Mock {
```
Comment (Stebalien) [inline internal/clock/clock.go:17]: Is this used?

File: manifest/manifest_sender.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/filecoin-project/go-f3/internal/clock"
```
Comment (Stebalien) [inline manifest/manifest_sender.go:9]: grouping

**Review by Stebalien — APPROVED** [review]
My one gripe is that we usually take the clock in the constructor, except when we don't have a constructor and we take it in `Start`. But I can live with that.

File: internal/clock/clock.go (hunk 3 above)
Comment (Kubuxu) [reply internal/clock/clock.go:17]: Yes in polling tests

Comment (Stebalien) [reply internal/clock/clock.go:19]: Ah, got it.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
Welp. That looks like a mock clock bug. We're blocked writing to the channel in `Tick`. I'm _guessing_ we should be selecting on default here...

But this case _shouldn't_ happen if we're resetting correctly. It looks like we're re-starting a timer before draining it.

**Comment by Stebalien** [comment]
I've forked to https://github.com/filecoin-project/go-clock/ (raul's version is behind anyways).

**Comment by Stebalien** [comment]
https://github.com/filecoin-project/go-clock/pull/1

**Comment by Stebalien** [comment]
> But this case shouldn't happen if we're resetting correctly. It looks like we're re-starting a timer before draining it.

That was due to a second bug. Fixed in the same PR.

**Pull request merged**